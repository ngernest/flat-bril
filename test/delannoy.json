{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "dest": "v1",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v2",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "ans",
          "funcs": [
            "count_path"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "ans"
          ],
          "dest": "v3",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v3"
          ],
          "op": "print"
        },
        {
          "dest": "v4",
          "op": "const",
          "type": "int",
          "value": 0
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "y",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x"
          ],
          "dest": "v1",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v3"
          ],
          "labels": [
            "then.0",
            "else.0"
          ],
          "op": "br"
        },
        {
          "label": "then.0"
        },
        {
          "dest": "v4",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v4"
          ],
          "op": "ret"
        },
        {
          "label": "else.0"
        },
        {
          "args": [
            "y"
          ],
          "dest": "v6",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v7",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v6",
            "v7"
          ],
          "dest": "v8",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v8"
          ],
          "labels": [
            "then.5",
            "else.5"
          ],
          "op": "br"
        },
        {
          "label": "then.5"
        },
        {
          "dest": "v9",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v9"
          ],
          "op": "ret"
        },
        {
          "label": "else.5"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v10",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v11",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v10",
            "v11"
          ],
          "dest": "v12",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "y"
          ],
          "dest": "v13",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v12",
            "v13"
          ],
          "dest": "v14",
          "funcs": [
            "count_path"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v15",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "y"
          ],
          "dest": "v16",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v17",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v16",
            "v17"
          ],
          "dest": "v18",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v15",
            "v18"
          ],
          "dest": "v19",
          "funcs": [
            "count_path"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v14",
            "v19"
          ],
          "dest": "v20",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v21",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v22",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v21",
            "v22"
          ],
          "dest": "v23",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "y"
          ],
          "dest": "v24",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v25",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v24",
            "v25"
          ],
          "dest": "v26",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v23",
            "v26"
          ],
          "dest": "v27",
          "funcs": [
            "count_path"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v20",
            "v27"
          ],
          "dest": "v28",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v28"
          ],
          "op": "ret"
        }
      ],
      "name": "count_path",
      "type": "int"
    }
  ]
}
