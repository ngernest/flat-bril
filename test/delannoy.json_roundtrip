{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "ans",
          "funcs": [
            "count_path"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "ans"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v3"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v4",
          "op": "const",
          "type": "int",
          "value": 0
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "y",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v3"
          ],
          "funcs": [],
          "labels": [
            "then.0",
            "else.0"
          ],
          "op": "br"
        },
        {
          "label": "then.0"
        },
        {
          "dest": "v4",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v4"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else.0"
        },
        {
          "args": [
            "y"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v7",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v6",
            "v7"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v8"
          ],
          "funcs": [],
          "labels": [
            "then.5",
            "else.5"
          ],
          "op": "br"
        },
        {
          "label": "then.5"
        },
        {
          "dest": "v9",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v9"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else.5"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v11",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v10",
            "v11"
          ],
          "dest": "v12",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "y"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v12",
            "v13"
          ],
          "dest": "v14",
          "funcs": [
            "count_path"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v15",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "y"
          ],
          "dest": "v16",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v17",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v16",
            "v17"
          ],
          "dest": "v18",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v15",
            "v18"
          ],
          "dest": "v19",
          "funcs": [
            "count_path"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v14",
            "v19"
          ],
          "dest": "v20",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v21",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v22",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v21",
            "v22"
          ],
          "dest": "v23",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "y"
          ],
          "dest": "v24",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v25",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v24",
            "v25"
          ],
          "dest": "v26",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v23",
            "v26"
          ],
          "dest": "v27",
          "funcs": [
            "count_path"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v20",
            "v27"
          ],
          "dest": "v28",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v28"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "count_path",
      "type": "int"
    }
  ]
}
