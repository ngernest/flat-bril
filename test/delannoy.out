@main
	op: id   		dest: v1		type: int		args: ["n"]	
	op: id   		dest: v2		type: int		args: ["n"]	
	op: call 		dest: ans		type: int		args: ["v1", "v2"]		funcs: ["count_path"]
	op: id   		dest: v3		type: int		args: ["ans"]	
	op: print		args: ["v3"]	
	op: const		dest: v4		type: int		value: 0	
@count_path
	op: id   		dest: v1		type: int		args: ["x"]	
	op: const		dest: v2		type: int		value: 0	
	op: eq   		dest: v3		type: bool		args: ["v1", "v2"]	
	op: br   		args: ["v3"]		labels: ["then.0", "else.0"]
.then.0
	op: const		dest: v4		type: int		value: 1	
	op: ret  		args: ["v4"]	
.else.0
	op: id   		dest: v6		type: int		args: ["y"]	
	op: const		dest: v7		type: int		value: 0	
	op: eq   		dest: v8		type: bool		args: ["v6", "v7"]	
	op: br   		args: ["v8"]		labels: ["then.5", "else.5"]
.then.5
	op: const		dest: v9		type: int		value: 1	
	op: ret  		args: ["v9"]	
.else.5
	op: id   		dest: v10		type: int		args: ["x"]	
	op: const		dest: v11		type: int		value: 1	
	op: sub  		dest: v12		type: int		args: ["v10", "v11"]	
	op: id   		dest: v13		type: int		args: ["y"]	
	op: call 		dest: v14		type: int		args: ["v12", "v13"]		funcs: ["count_path"]
	op: id   		dest: v15		type: int		args: ["x"]	
	op: id   		dest: v16		type: int		args: ["y"]	
	op: const		dest: v17		type: int		value: 1	
	op: sub  		dest: v18		type: int		args: ["v16", "v17"]	
	op: call 		dest: v19		type: int		args: ["v15", "v18"]		funcs: ["count_path"]
	op: add  		dest: v20		type: int		args: ["v14", "v19"]	
	op: id   		dest: v21		type: int		args: ["x"]	
	op: const		dest: v22		type: int		value: 1	
	op: sub  		dest: v23		type: int		args: ["v21", "v22"]	
	op: id   		dest: v24		type: int		args: ["y"]	
	op: const		dest: v25		type: int		value: 1	
	op: sub  		dest: v26		type: int		args: ["v24", "v25"]	
	op: call 		dest: v27		type: int		args: ["v23", "v26"]		funcs: ["count_path"]
	op: add  		dest: v28		type: int		args: ["v20", "v27"]	
	op: ret  		args: ["v28"]	
